src/linear_algebra/basis.lean

```diff
@@ -263,48 +275,66 @@ lemma range_reindex : set.range (b.reindex e) = set.range b :=
 by rw [coe_reindex, range_reindex']
 
 /-- `b.reindex_range` is a basis indexed by `range b`, the basis vectors themselves. -/
-def reindex_range [nontrivial R] : basis (range b) R M :=
-b.reindex (equiv.of_injective b b.injective)
+def reindex_range : basis (range b) R M :=
+begin
```

Comment:
I am not sure it is OK to use tactic mode in a `def`.

Reply:
Hmm, I'm not sure this is a good idea. It would be nice if we can avoid `by_cases` in this definition.

Reply:
Let me see if there is a better proof.

Reply:
I think this is fine. The lemmas below show that you can still reason about the value of this definition. You could also use `or.by_cases` in term mode, but you will probably end up with the same / a similar term.

Reply:
Personally, I would prefer `if` for definitions:
```lean
if h : nontrivial R then
  by exactI
  ...
else
  by haveI : subsingleton R := ...; exact
  ...
```

Reply:
I like the `if`! I have the impression we have to reason by cases, in a way or another: the two proofs are really different. Indeed if `R` is trivial this is true just because anything is a basis (and the rank is not well defined!).